### DISCUSSION TITLE
`text-to-sdf` expression

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-style-spec#145] [Ideas]
I wonder if it would make sense to add a style spec expression called `text-to-sdf` which takes a string, renders it in a canvas and turns the image into an sdf.

A use-case would be:

```json
"text-field": "Bangladesh",
"icon-image": ["text-to-sdf", "বাংলাদেশ"]
```

With this, we could do stuff like this:

<a href="https://wipfli.github.io/maplibre-canvas-labels/offscreen#map=5.97/23.64/90.874/12">
<img src="https://user-images.githubusercontent.com/53421382/234887223-babb6df7-2886-4d44-b3ca-7f4a33d03adb.png" width=650>
</a>

What do you think?

**Comment by HarelM** [comment]
What problem does this solve? If the problem is with the current support of text, then it would probably have the same problems.
A use case I can think of, which I tend to think should have a better solution, is for text that follows streams, where the letters get messed up instead of following the steam curves. So this feature will allow keeping the text straight (a better implementation would be to following a straight line between two points on the stream basically).

**Reply by wipfli** [reply]
Using text rendering in a canvas means you can use the browser to render the text. So you can basically render all languages/writing systems with this solution...

**Reply by HarelM** [reply]
Sounds like a solution for a specific platform, isn't it?

**Reply by wipfli** [reply]
Canvas is web specific, right. Let me see if mobile operating systems have something similar which lets you turn text to pixels...

**Reply by 1ec5** [reply]
Yes, every platform has an analogue. For example, on iOS, a [`UIGraphicsImageRenderer`](https://developer.apple.com/documentation/uikit/uigraphicsimagerenderer/) works very much like a canvas on the Web. You’d use Core Text APIs to render text using system fonts, just as gl-native does when generating CJK SDFs.

But @HarelM has a good point that this approach fundamentally only works for point features where you can reasonably hard-code the image’s dimensions at design time. Line-placed labels are important in any map that provides more detail than a city locator map. Another downside is the inability to reuse rasterized labels, leading to bloat in the internal spritesheet. And the drawing function doesn’t have enough context to rotate or tilt labels if the label needs to align to the map instead of the viewport.

`styleimagemissing` is a workable solution for rendering route shields because they are inherently point features that align with the viewport. But the issue of spritesheet bloat still comes up anyways.

In my opinion, the lack of built-in complex text support is problematic enough that any solution, however limited, deserves consideration for the benefit of the affected language communities. However, this feels to me like a workaround to be published as a plugin or example rather than a well-integrated solution.

**Reply by wipfli** [reply]
Yes maybe it would be best if I started with a plugin! I have never written one, do you have an example I could follow?

**Reply by 1ec5** [reply]
[mapbox-gl-language](https://github.com/mapbox/mapbox-gl-language/) is a good example of a plugin that’s all about runtime styling.

**Comment by wipfli** [comment]
Using a canvas in web browsers would mean that the text does not look identical on all devices. Is this a requirement which we can drop for the sake of being able to render at least something in currently non-supported languages?

**Comment by wipfli** [comment]
Collisions are broken, and long labels look not great, but still, here is an example with street names in Bengali:

https://wipfli.github.io/maplibre-canvas-labels/offscreen#map=14.98/23.75498/90.38573

![image](https://user-images.githubusercontent.com/53421382/234952201-01564d37-3088-487b-abba-6dd11a6073b1.png)

**Reply by wipfli** [reply]
I think the proper way to do canvas labels along a line is to support sdf images in formatted expressions. Currently, only normal images are supported there...

**Reply by wipfli** [reply]
But `text-to-sdf` would also be needed in that case.

**Reply by 1ec5** [reply]
The hard part is curving a label along a curved line. The rendering engine doesn’t share any context about the line geometry with the `styleimagemissing` handler (because an image isn’t inherently tied to a particular feature). You could imagine rendering each [extended grapheme cluster](https://unicode.org/reports/tr29/#Table_Sample_Grapheme_Clusters) as a separate image and stringing the images together as a formatted expression. But while there are JavaScript libraries for splitting a string into extended grapheme clusters, there’s nothing like that in the expression language, which is currently responsible for specifying which images make up a formatted expression. Even splitting by a static string isn’t supported (maplibre/maplibre-gl-js#2064), let alone by a Unicode-aware regular expression, let alone by extended grapheme clusters.

**Reply by wipfli** [reply]
I updated my tiles to include road names.

Feel free to have  a look at this global demo with street names:

https://wipfli.github.io/maplibre-canvas-labels/offscreen#map=15.77/11.566654/104.922352

![image](https://user-images.githubusercontent.com/53421382/235134572-b610121d-74f6-46ec-bcac-2ba5b09beab1.png)

**Reply by wipfli** [reply]
Thanks for pointing to the term "grapheme clusters". Interesting thing... Do you think libraries like this one are any good? https://github.com/orling/grapheme-splitter

**Reply by 1ec5** [reply]
It claims to implement UAX #29, which is a good sign. Incidentally, ECMAScript specifies an [`Intl.Segmenter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter) that also implements this standard, segmenting by grapheme cluster by default (with options for word and sentence splitting). There are analogues on other platforms, such as [iOS/macOS](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Strings/Articles/stringsClusters.html). Ideally, GL JS and gl-native would use these facilities for more robust character shaping along curved lines out of the box, but they could be useful to you as part of the workaround. Note that `Intl.Segmenter` is very new and [not yet supported on Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1423593).

**Reply by wipfli** [reply]
Uh, nice, thanks for the Intl.Segmenter hint! It seems to actually work better than this javascript library I posted above:

According to the [grapheme splitter readme](https://github.com/orling/grapheme-splitter), the js library splits

अनुच्छेद into अ + नु + च् + छे + द

whereas I find that IntlSegtmenter splits

अनुच्छेद into अ + नु + च्छे + द

To me, the Intl.Segmenter splitting looks much closer to the original...

**Reply by wipfli** [reply]
If we do this in `glyph_manager.ts`:

```js
const letters = ['ཡུ', 'ལ་', 'ཤུ', 'ལ་', 'བོ', 'ད', '་རི', 'ག', 'ས', '་ར', 'ང', '་སྐྱོ', 'ང', '་ཁུལ།'];
const char = tinySDF.draw(letters[Math.floor(Math.random() * letters.length)]);
```

And always use the TinySDF codepath with:

```js
  _doesCharSupportLocalGlyph(id: number): boolean {
        return true;
        /* eslint-disable new-cap */
        return !!this.localIdeographFontFamily &&
            (isChar['CJK Unified Ideographs'](id) ||
                isChar['Hangul Syllables'](id) ||
                isChar['Hiragana'](id) ||
                isChar['Katakana'](id));
        /* eslint-enable new-cap */
    }
```

Then we get this:

![image](https://user-images.githubusercontent.com/53421382/235296437-c9fe1bff-e02a-4c54-9dd6-6e68fcfecad7.png)

**Reply by wipfli** [reply]
So I think we should do the following:

1. Stop using the unicode number (type number) for indexing glyphs, instead index by character (type string).
2. Do not request glyphs by codepoint, instead request them by grapheme cluster. 

I think if I can get step 1 working, then step 2 will be trivial and we will have pretty much full writing system support.

**Reply by 1ec5** [reply]
Grapheme cluster segmentation is independent of the font, so it doesn’t account for ligatures that exist in some fonts but not others. I don’t know to what extent this is an issue for currently unsupported scripts, but it would potentially be an issue for supported scripts like Latin. For example, some fonts would display “fi” as the ligature “ﬁ” at certain sizes, but this approach would prevent that from happening because the code handling segmentation wouldn’t know anything about the font.

This would still be an improvement over the current behavior with respect to Indic scripts. But if MapLibre officially adopts a new server API based on this approach to replace the existing fontstack API, then it might make it even more difficult to implement something more correct around Harfbuzz in the future. If you can find a way to retrofit this approach into a purely local codepath while keeping the server-rendered codepath intact, that would minimize disruption.

**Reply by ramSeraph** [reply]
> Uh, nice, thanks for the Intl.Segmenter hint! It seems to actually work better than this javascript library I posted above:
> 
> According to the [grapheme splitter readme](https://github.com/orling/grapheme-splitter), the js library splits
> 
> अनुच्छेद into अ + नु + च् + छे + द
> 
> whereas I find that IntlSegtmenter splits
> 
> अनुच्छेद into अ + नु + च्छे + द
> 
> To me, the Intl.Segmenter splitting looks much closer to the original...

This looks quite good. I hope firefox gets the Intl.Segmenter soon.

**Reply by wipfli** [reply]
https://github.com/wipfli/index-by-grapheme

**Reply by wipfli** [reply]
We can also segement by word:
```diff
         const segmenter = new Intl.Segmenter(
-             'en', {granularity: 'grapheme'}
+             'en', {granularity: 'word'}
         );
```

Here is a demo: https://wipfli.github.io/index-by-grapheme/by-word/#map=4.81/16.04/82.31&language=te

**Reply by wipfli** [reply]
For Burmese, one might want to use `granularity: 'sentence'`. Maybe we can let the user decide which granularity to use...

**Reply by 1ec5** [reply]
Word and sentence segmentation is even more dependent on the text’s locale, which this code doesn’t have access to. Anything beyond grapheme cluster would behave similar to not segmenting at all (what you had tried originally): it would look disjointed and bloat the glyph texture.

[<img width="317" alt="Uppuguda Station Road" src="https://user-images.githubusercontent.com/1231218/235343798-8a3f34f3-b5cf-4fc0-90f9-2eb4452a8f91.png">](https://wipfli.github.io/index-by-grapheme/by-word/#map=15.02/17.34329/78.47948&language=te) [<img width="372" alt="Bhuvaneshwari Road" src="https://user-images.githubusercontent.com/1231218/235343921-83c9f46f-a1b1-4a16-b341-bd4f280240a9.png">](https://wipfli.github.io/index-by-grapheme/by-word/#map=15.73/9.975995/76.309891&language=te)

**Comment by wipfli** [comment]
New idea: We could let the user decide how to segment text. The user could insert some sort of special character at the locations where text can be segmented. 

Examples:

* Siem Reap (Khmer):  
  * សៀមរាប
  * សៀ + ម + រា + ប
* Dubai (Arabic): 
  * Dubai دبي  
  * Dubai د + بي
* Crescent Lake (Bengali):
  * ক্রিসেন্ট লেক
  * ক্রিসে + ন্ট +   + লে + ক
* Bogale (Burmese):
  * ဘိုကလေးမြို့
  * ဘို + က + လေး + မြို့

**Reply by wipfli** [reply]
I don't know how to do the segmentation for every language correctly, but I think there must be ways. And if we do the segmentation at tile-generation time, we can use as many libraries and as much compute as we want...
